### PR TITLE
James 2065 mailet attribute collection to mime headers 

### PR DESCRIPTION
**Pull request by chibenwa**
Rebase of of https://github.com/linagora/james-project/pull/857

Added value:
 - Improve assertJ assertions
 - Java-8
 - Better resiliance with casts
 - Wider type: collection instead of List
 - Apply logging best practices

### CODE REVIEW

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/MappingArgument.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        return Splitter.on(',').splitToList(mapping)
+            .stream()
+            .map(MappingArgument::parseKeyValue)
+            .collect(Guavate.toImmutableMap(Pair::getLeft, Pair::getRight));
+    }
+
+    private static Pair<String, String> parseKeyValue(String keyValue) {
+        List<String> pair = Splitter.on(';').trimResults().splitToList(keyValue);
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/MappingArgument.java]: this trim here is weird, and you don't have tests for it AFAIK

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/MappingArgumentTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Test
+    public void parseShouldFailIfCalledWithBadSyntax() throws MessagingException {
+        expectedException.expect(IllegalArgumentException.class);
+        MappingArgument.parse("key1;value1,key2");
+    }
+
+    @Test
+    public void parseShouldFailIfCalledWithBadSyntax2() throws MessagingException {
```
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/MappingArgumentTest.java]: 2 ?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeaders.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        attributeNameToHeader = MappingArgument.parse(simpleMappings);
+    }
+
+    @Override
+    public void service(Mail mail) {
+        try {
+            MimeMessage message = mail.getMessage();
+            attributeNameToHeader.entrySet()
+                .forEach(entry -> addAttributeToHeaders(mail, message, entry));
+            message.saveChanges();
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeaders.java]: do we really need to call save ?

Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeaders.java]: Apparently yes...

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/MappingArgument.java (hunk 1 above)
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/MappingArgument.java]: I will add them

File: mailet/standard/pom.xml [hunk 4]
```diff
@@ -55,6 +55,11 @@
             <version>${mime4j.version}</version>
         </dependency>
         <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>james-server-util-java8</artifactId>
+            <version>${project.version}</version>
```
Comment (mbaechler) [inline mailet/standard/pom.xml]: you should not need the version

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+        assertThat(mail.getMessage().getHeader(HEADER_NAME1))
+            .containsExactlyElementsOf(MAIL_ATTRIBUTE_VALUE1);
+
+        assertThat(mail.getMessage().getHeader(HEADER_NAME2))
+            .containsExactlyElementsOf(MAIL_ATTRIBUTE_VALUE2);
+    }
+
+    @Test
+    public void shouldNotRemovePreviousAttributeValueWhenAttributeAlreadyPresent() throws MessagingException {
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java]: s/Attribute/Header/g ?

File: mailet/standard/pom.xml (hunk 4 above)
Comment (chibenwa) [inline mailet/standard/pom.xml]: True, but I'm a bit fed up with merge conflicts with pom pr...

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+            .build();
+
+        mailet.service(mail);
+
+        assertThat(mail.getMessage().getHeader(HEADER_NAME1))
+            .containsOnly(VALUE_1_1, VALUE_1_2, firstValue);
+    }
+
+    @Test
+    public void test() throws MessagingException {
```
Comment (quynhn) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java]: Naming?

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        mailet.service(mail);
+
+        assertThat(mail.getMessage().getHeader(HEADER_NAME1)).isNull();
+        assertThat(mail.getMessage().getHeader(HEADER_NAME2))
+            .containsExactlyElementsOf(MAIL_ATTRIBUTE_VALUE2);
+    }
+
+
+    @Test
+    public void test2() throws MessagingException {
```
Comment (quynhn) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java]: idem

Comment (quynhn) [reply mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java]: Saw

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

docker ldap failure

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java (hunk 5 above)
Comment (rouazana) [reply mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java]: not fixed?

**Comment by chibenwa** [comment]
```
Results :

Tests in error: 
  HBaseUidAndModSeqProviderTest.testHighestModSeq » NoClassDefFound Could not in...
  HBaseUidAndModSeqProviderTest.testLastUid » ExceptionInInitializer
  HBaseUidAndModSeqProviderTest.testNextModSeq » NoClassDefFound Could not initi...
  HBaseUidAndModSeqProviderTest.testNextUid » NoClassDefFound Could not initiali...
```

test this please

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java (hunk 5 above)
Comment (rouazana) [reply mailet/standard/src/test/java/org/apache/james/transport/mailets/MailAttributesListToMimeHeadersTest.java]: fixed now :)

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**